wgpu/src/image.rs

```diff
@@ -244,44 +282,76 @@ impl Pipeline {
             std::mem::size_of::<Uniforms>() as u64,
         );
 
-        // TODO: Batch draw calls using a texture atlas
-        // Guillotière[1] by @nical can help us a lot here.
-        //
-        // [1]: https://github.com/nical/guillotiere
-        for image in instances {
-            let uploaded_texture = match &image.handle {
-                Handle::Raster(handle) => {
-                    let mut cache = self.raster_cache.borrow_mut();
-                    let memory = cache.load(&handle);
+        let mut recs = HashMap::new();
 
-                    memory.upload(device, encoder, &self.texture_layout)
+        for (index, image) in instances.iter().enumerate() {
+            match &image.handle {
+                Handle::Raster(handle) => {
+                    let mut raster_cache = self.raster_cache.borrow_mut();
+
+                    if let Memory::Device(allocation) = raster_cache.upload(
+                        handle,
+                        device,
+                        encoder,
+                        &mut self.allocator,
+                        &mut self.atlas
+                    ) {
+                        let rec = allocation.rectangle;
+
+                        let _ = recs.insert(index, rec);
+                    }
                 }
                 Handle::Vector(_handle) => {
                     #[cfg(feature = "svg")]
                     {
-                        let mut cache = self.vector_cache.borrow_mut();
+                        let mut vector_cache = self.vector_cache.borrow_mut();
 
-                        cache.upload(
+                        // Upload rasterized svg to texture atlas
+                        if let Some(allocation) = vector_cache.upload(
                             _handle,
                             image.scale,
                             _scale,
                             device,
                             encoder,
-                            &self.texture_layout,
-                        )
-                    }
+                            &mut self.allocator,
+                            &mut self.atlas,
+                        ) {
+                            let rec = allocation.rectangle;
 
-                    #[cfg(not(feature = "svg"))]
-                    None
+                            let _ = recs.insert(index, rec);
+                        }
+                    }
                 }
-            };
+            }
+        }
+
+        let atlas_width = self.allocator.size().width as f32;
+        let atlas_height = self.allocator.size().height as f32;
+
+        let texture = device.create_bind_group(&wgpu::BindGroupDescriptor {
```

Comment:
Do we need to create a texture view and a bind group every frame?

Reply:
No